### PR TITLE
Ioc refactor

### PR DESCRIPTION
**Pull request by petermartigny**
- Refactor the basic trainer in an IoC pattern, with a single registry for every registrable classes, allowing for maximum customization
- Separate the example experiments from the library
- Adapt the examples to the new logic
- Set cuda as optional in the config file

### CODE REVIEW

File: experiments/cbow.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+from transfer_nlp.embeddings.embeddings import Embedding
+from transfer_nlp.loaders.loaders import DatasetSplits, DataFrameDataset, DatasetHyperParams
+from transfer_nlp.loaders.vectorizers import VectorizerNew
+from transfer_nlp.loaders.vocabulary import CBOWVocabulary
+from transfer_nlp.plugins.config import register_plugin
+from transfer_nlp.plugins.helpers import ObjectHyperParams
+from transfer_nlp.plugins.predictors import Predictor, PredictorHyperParams
+
+
+name = 'transfer_nlp.experiments.cbow'
```
Comment (kireet) [inline experiments/cbow.py]: https://stackoverflow.com/a/603460

i guess the proper way to do this is by using `__name__`

File: experiments/cbow.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+logging.getLogger(name).setLevel(level=logging.INFO)
+logger = logging.getLogger(name)
+
+# Vectorizer
+@register_plugin
+class CBOWVectorizer(VectorizerNew):
```
Comment (kireet) [inline experiments/cbow.py]: `VectorizerNew`? maybe we should rename this before merging

File: experiments/cbow.json [hunk 3]
```diff
@@ -0,0 +1,79 @@
+{
+  "data_file": "HOME/work/experiments/nlp/data/books/frankenstein_with_splits.csv",
+  "model_state_file": "modelCBOW.pth",
```
Comment (kireet) [inline experiments/cbow.json]: are all these string properties used? maybe we should clean out the unused ones...after merging i could add a warning message for unused config params. that would make it easier. we could postpone cleanup til then.

File: experiments/cbow.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    def __init__(self, predictor_hyper_params: PredictorHyperParams):
+        super().__init__(predictor_hyper_params=predictor_hyper_params)
+
+    def json_to_data(self, input_json: Dict) -> Dict:
+        return {
+            'x_in': torch.LongTensor([self.vectorizer.vectorize(context=input_string) for input_string in input_json['inputs']])}
+
+    def output_to_json(self, outputs: List[Dict[str, Any]]) -> Dict[str, Any]:
+        return {
+            "outputs": outputs}
```
Comment (kireet) [inline experiments/cbow.py]: formatting?

File: experiments/cbow.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    """
+    Toy example: we want to make predictions on inputs of the form {"inputs": ["hello world", "foo", "bar"]}
+    """
+
+    def __init__(self, predictor_hyper_params: PredictorHyperParams):
+        super().__init__(predictor_hyper_params=predictor_hyper_params)
+
+    def json_to_data(self, input_json: Dict) -> Dict:
+        return {
+            'x_in': torch.LongTensor([self.vectorizer.vectorize(context=input_string) for input_string in input_json['inputs']])}
```
Comment (kireet) [inline experiments/cbow.py]: formatting?

File: experiments/training.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+from experiments.cbow import *
+
+name = 'transfer_nlp.experiments.training'
+logging.getLogger(name).setLevel(level=logging.INFO)
+logger = logging.getLogger(name)
+
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.INFO)
+
+    surname_paths = ['/Users/petermartigny/Documents/PycharmProjects/transfer-nlp/experiments/mlp.json',
```
Comment (kireet) [inline experiments/training.py]: shouldn't check in user directories into git. probably not pycharm either. 
maybe use some very generic convention like `str(Path.home() / 'work/transfer_nlp'`?

File: transfer_nlp/plugins/helpers.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    a model input size that depends on the data set composition.
+    """
+
+    def __init__(self):
+        self.input_dim: int = None
+        self.output_dim: int = None
+
+
+@register_plugin
+class ModelParams:
```
Comment (kireet) [inline transfer_nlp/plugins/helpers.py]: i was thinking this class should be renamed `TrainableParameters` or something? it makes it more clear its dealing with weights to optimize but still retains the parameters terminology from pytorch...

File: transfer_nlp/plugins/predictors.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+@register_plugin
+class PredictorHyperParams(ObjectHyperParams):
+
+    def __init__(self, trainer: BasicTrainer):
+        super().__init__()
+        self.vectorizer = trainer.dataset_splits.vectorizer
+        self.model = trainer.model
+
+
+class Predictor:
```
Comment (kireet) [inline transfer_nlp/plugins/predictors.py]: should this be an ABC class? I would expect a `predict` method from a class called `Predictor`. Also given that usually pytorch forward methods leave off the final activation function to optimize training, maybe the predictor could take in a function to apply to the output? or that's too much complexity to be worth the small gain?

File: transfer_nlp/plugins/trainers.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+            metric.attach(engine, f'{str(metric.__class__.__name__)}')
+
+        return engine
+
+    def train(self):
+        self.trainer.run(self.dataset_splits.train_data_loader(), max_epochs=self.num_epochs)
+
+# Things available to add for custom trainers
+
+# We show here how to add some events: tensorboard logs!
```
Comment (kireet) [inline transfer_nlp/plugins/trainers.py]: this stuff i left as TODOs to add in once we both agreed on the basic structure...we can add it back in future PRs?

File: experiments/cbow.py (hunk 2 above)
Comment (petermartigny) [reply experiments/cbow.py]: I changed to `Vectorizer`

File: transfer_nlp/plugins/predictors.py (hunk 8 above)
Comment (petermartigny) [reply transfer_nlp/plugins/predictors.py]: Yes, this should be an ABC class,
for now the `infer` method does just what the `infer` method from the nn.Module does, and the `decode` method is an abstract method that one has to implement depending on the use case

File: experiments/training.py (hunk 6 above)
Comment (petermartigny) [reply experiments/training.py]: This file should be run from the `experiments/` folder so I changed them to `"./experiment_name"`

File: experiments/cbow.json (hunk 3 above)
Comment (petermartigny) [reply experiments/cbow.json]: we don't use model checkpointing in the `BasicTrainer` so `model_state_file` can be omitted for now

**Comment by petermartigny** [comment]
IoC Refactoring

**Pull request merged**